README.md
# Octicons gem

[![Gem version](https://img.shields.io/gem/v/octicons.svg)](https://rubygems.org/gems/octicons)
[![Build Status](https://travis-ci.org/primer/octicons_gem.svg?branch=master)](https://travis-ci.org/primer/octicons_gem)

> Octicons gem to distribute octicons svg

## Install

Add this to your `Gemfile`

```rb
gem 'octicons'
```

Then `bundle install`.

## Usage

```rb
require 'octicons'
icon = Octicons::Octicon.new("x")
icon.to_svg
# <svg class="octicon octicon-x" viewBox="0 0 16 16" width="16" height="16" version="1.1" "aria-hidden"="true"><path d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48z"></path></svg>
```

## Documentation

The `Octicon` class takes two arguments. The first is the symbol of the icon, and the second is a hash of arguments representing html attributes

#### `symbol` _(required)_

This is the name of the octicon you want to use. For example `alert`. [Full list of icons][octicons-docs]

#### Options

* `:height` - When setting the height to a number, the icon will scale to that size. For example, passing `32`, will calculate the width based on the icon's natural size.
* `:width` - When setting the width to a number, the icon will scale to that size. For example, passing `32`, will calculate the width based on the icon's natural size.

If both `:width, :height` are passed into the options hash, then the icon will be sized exactly at those dimensions.

#### Attributes

Once initialized, you can read a few properties from the icon.

##### `symbol`

Returns the string of the symbol name

```rb
icon = Octicons::Octicon.new("x")
icon.symbol
# "x"
```

##### `path`

Path returns the string representation of the path of the icon.

```rb
icon = Octicons::Octicon.new("x")
icon.path
# <path d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48z"></path>
```

##### `options`

This is a hash of all the `options` that will be added to the output tag.

```rb
icon = Octicons::Octicon.new("x")
icon.options
# {:class=>"octicon octicon-x", :viewBox=>"0 0 12 16", :version=>"1.1", :width=>12, :height=>16, :"aria-hidden"=>"true"}
```

##### `width`

Width is the icon's true width. Based on the svg view box width. _Note, this doesn't change if you scale it up with size options, it only is the natural width of the icon_

##### `height`

Height is the icon's true height. Based on the svg view box height. _Note, this doesn't change if you scale it up with size options, it only is the natural height of the icon_

#### Methods

##### `to_svg`

Returns a string of the svg tag

```rb
icon = Octicons::Octicon.new("x")
icon.to_svg
# <svg class="octicon octicon-x" viewBox="0 0 16 16" width="16" height="16" version="1.1" "aria-hidden"="true"><path d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48z"></path></svg>
```

##### `keywords`

Returns an array of keywords for the icon. The data [comes from the octicons repository](https://github.com/primer/octicons/blob/master/lib/keywords.json). Consider contributing more aliases for the icons.

```rb
icon = Octicons::Octicon.new("x")
icon.keywords
# ["remove", "close", "delete"]
```

## Publishing

If you have access to publish this repository, these are the steps to publishing. If you need access, contact [#design-systems](https://github.slack.com/archives/design-systems).

**Before publishing** This repository relies on the data from [octicons][]. To update to the most recent version, you'll need to run `npm run update`

1. Update the [CHANGELOG.md](./CHANGELOG.md) with relevant version number and any updates made to the repository.
2. Update the version in [version.rb](https://github.com/primer/octicons_gem/blob/master/lib/octicons/version.rb) using the relevant version. The versioning is [semver](http://semver.org/), so version appropriately based on what has changed.
3. `npm version <newversion>` Use the same version that you added in step 2.
4. `npm run ship` This will build the gem and publish it to rubygems.
5. `git push && git push --tags` Push all these changes to origin.

## License

(c) 2012-2016 GitHub, Inc.

When using the GitHub logos, be sure to follow the [GitHub logo guidelines](https://github.com/logos).

_Font License:_ [SIL OFL 1.1](http://scripts.sil.org/OFL)  
Applies to all font files and SVG files

_Code License:_ [MIT](./LICENSE)  
Applies to all other files

[octicons]: https://github.com/primer/octicons
[octicons-docs]: https://octicons.github.com/


CHANGELOG.md
# HEAD

# 4.0.0

- Updating to new version of [octicons@6.0.0](https://github.com/primer/octicons/blob/master/CHANGELOG.md#600)

# 3.0.1

- Fixes: project icon wasn't rendering properly

# 3.0.0

- Added: `to_svg_use` which will output a `<svg><use>` tag for svg spritesheets
- Added: Ability to output the spritesheet with `Octicons.sprite_sheet`
- Updating to new version of [octicons@5.0.0](https://github.com/primer/octicons/blob/master/CHANGELOG.md#500)
- Removed: api endpoints for `decimal`, `hexadecimal`, and `character`
- Removed: svg directory from `/lib/` this isn't necessary anymore with the updated `data.json` from 5.0.0
- Removed: `Octicons::KEYWORDS` and `Octicons::CODEPOINTS`, Codepoints is deprecated and Keywords can be accessed via `Octicons::OCTICON_SYMBOLS`

# 2.1.0

- Updating to new version of [octicons@4.3.0](https://github.com/primer/octicons/blob/master/CHANGELOG.md#430)

# 2.0.0

- Adding: attr_reader for `:path, :options, :width, :height` https://github.com/primer/octicons_gem/pull/6
- Adding: `keywords.json` and `codepoints.json` methods to the octicon object. Making them available via the api https://github.com/primer/octicons_gem/pull/7
- Deprecating: the use of `:size` for "large" or a number, instead use `:height` https://github.com/primer/octicons_gem/pull/8
- Deprecate: the use of `:symbol`, instead the first argument is the symbol https://github.com/primer/octicons_gem/pull/9

# 1.1.0

- Added new octicons version 4.1.0

# 1.0.0

- First release, it's an octicons gem.


test/helper.rb
require 'minitest/autorun'
require 'octicons'

def octicon(options)
  ::Octicons::Octicon.new(options)
end


test/octicon_test.rb
require_relative "./helper"

describe Octicons::Octicon do
  it "fails when the octicon doesn't exist" do
    assert_raises(RuntimeError) do
      octicon("octicon")
    end
  end

  it "initialize accepts a string for an icon" do
    icon = octicon("x")
    assert icon
  end

  it "initialize accepts a symbol for an icon" do
    icon = octicon(:x)
    assert icon
  end

  it "gets keywords for the icon" do
    icon = octicon("x")
    assert_equal ["remove", "close", "delete"], icon.keywords
  end

  it "the attributes are readable" do
    icon = octicon("x")
    assert icon.path
    assert icon.options
    assert_equal "x", icon.symbol
    assert_equal 12, icon.width
    assert_equal 16, icon.height
  end

  describe "codepoints" do
    it "computes the correct decimal codepoint" do
      icon = octicon("alert")
      assert_equal 61485, icon.decimal
    end

    it "computes the correct hexadecimal codepoint" do
      icon = octicon("alert")
      assert_equal "f02d", icon.hexadecimal
    end

    it "finds the character code" do
      icon = octicon("alert")
      assert_equal "", icon.character
    end
  end

  describe "viewBox" do
    it "always has a viewBox" do
      icon = octicon("x")
      assert_includes icon.to_svg, "viewBox=\"0 0 12 16\""
    end
  end

  describe "html_attributes" do
    it "includes other html attributes" do
      icon = octicon("x", :foo => "bar", :disabled => "true")
      assert_includes icon.to_svg, "disabled=\"true\""
      assert_includes icon.to_svg, "foo=\"bar\""
    end
  end

  describe "classes" do
    it "includes classes passed in" do
      icon = octicon("x", :class => "text-closed")
      assert_includes icon.to_svg, "class=\"octicon octicon-x text-closed\""
    end
  end

  describe "size" do
    it "always has width and height" do
      icon = octicon("x")
      assert_includes icon.to_svg, "height=\"16\""
      assert_includes icon.to_svg, "width=\"12\""
    end

    it "converts number string height to integer" do
      icon = octicon("x", :height => "60")
      assert_includes icon.to_svg, "height=\"60\""
      assert_includes icon.to_svg, "width=\"45\""
    end

    it "converts number height to integer" do
      icon = octicon("x", :height => 60)
      assert_includes icon.to_svg, "height=\"60\""
      assert_includes icon.to_svg, "width=\"45\""
    end

    it "converts number string width to integer" do
      icon = octicon("x", :width => "45")
      assert_includes icon.to_svg, "height=\"60\""
      assert_includes icon.to_svg, "width=\"45\""
    end

    it "converts number width to integer" do
      icon = octicon("x", :width => 45)
      assert_includes icon.to_svg, "height=\"60\""
      assert_includes icon.to_svg, "width=\"45\""
    end

    it "with height and width passed in" do
      icon = octicon("x", :width => 60, :height => 60)
      assert_includes icon.to_svg, "width=\"60\""
      assert_includes icon.to_svg, "height=\"60\""
    end
  end

  describe "a11y" do
    it "includes attributes" do
      icon = octicon("x", :'aria-label' => "Close")
      assert_includes icon.to_svg, "role=\"img\""
      assert_includes icon.to_svg, "aria-label=\"Close\""
    end

    it "has aria-hidden when no label is passed in" do
      icon = octicon("x")
      assert_includes icon.to_svg, "aria-hidden=\"true\""
    end
  end
end


test/octicons_test.rb
require_relative "./helper"

describe Octicons do

  def setup
    @icons_total = Dir[Octicons::OCTICONS_SVG_PATH].length
  end

  it "loads all icons on initialization" do
    refute_equal 0, Octicons::OCTICON_SYMBOLS.length
    assert_equal @icons_total, Octicons::OCTICON_SYMBOLS.length
  end

  it "loads all keywords" do
    refute_equal 0, Octicons::KEYWORDS.length
    assert_equal @icons_total, Octicons::KEYWORDS.length
  end

  it "loads all codepoints" do
    refute_equal 0, Octicons::CODEPOINTS.length
    assert_equal @icons_total, Octicons::CODEPOINTS.length
  end
end


lib/octicons/octicon.rb
module Octicons
  class Octicon

    attr_reader :path, :options, :width, :height, :symbol

    def initialize(symbol, options = {})
      @symbol = symbol.to_s
      if octicon = Octicons::OCTICON_SYMBOLS[@symbol]

        @path = octicon[:path]
        @width = octicon[:width]
        @height = octicon[:height]

        @options = options
        @options.merge!({
          :class   => classes,
          :viewBox => viewbox,
          :version => "1.1"
        })
        @options.merge!(size)
        @options.merge!(a11y)
      else
        raise "Couldn't find octicon symbol for #{@symbol.inspect}"
      end
    end

    # Returns an string representing a <svg> tag
    def to_svg
      "<svg #{html_attributes}>#{@path}</svg>"
    end

    # Returns an array of keywords similar to the icon
    def keywords
      Octicons::KEYWORDS[@symbol]["keywords"]
    end

    # Returns the decimal codepoint of the character
    def decimal
      Octicons::CODEPOINTS[@symbol]
    end

    # Returns the hexidecimal version of the character
    def hexadecimal
      decimal.to_s(16)
    end

    # Returns the unicode character
    def character
      [decimal].pack("U")
    end

    private

    def html_attributes
      attrs = ""
      @options.each { |attr, value| attrs += "#{attr}=\"#{value}\" " }
      attrs.strip
    end

    # add some accessibility features to svg
    def a11y
      accessible = {}

      if @options[:'aria-label'].nil?
        accessible[:'aria-hidden'] = "true"
      else
        accessible[:role] = "img"
      end

      accessible
    end

    # prepare the octicon class
    def classes
      "octicon octicon-#{@symbol} #{@options[:class]} ".strip
    end

    def viewbox
      "0 0 #{@width} #{@height}"
    end

    # determine the height and width of the octicon based on :size option
    def size
      size = {
        :width => @width,
        :height => @height
      }

      # Specific size
      unless @options[:width].nil? && @options[:height].nil?
        size[:width]  = @options[:width].nil?  ? calculate_width(@options[:height]) : @options[:width]
        size[:height] = @options[:height].nil? ? calculate_height(@options[:width]) : @options[:height]
      end

      size
    end

    def calculate_width(height)
      (height.to_i * @width) / @height
    end

    def calculate_height(width)
      (width.to_i * @height) / @width
    end
  end
end
